[pip][design] PIP-272: Add stateStorageConfig to WorkerConfig (#20455)
apache/pulsar: pip/pip-272.md

# Background knowledge

In Pulsar, a pulsar function support storing state, such as a `WordCount` function which stores the state of its counters.

```python
from pulsar import Function

class WordCount(Function):
    def process(self, item, context):
        for word in item.split():
            context.incr_counter(word, 1)
```

> Currently, Pulsar uses Bookkeeper as the default state storage interface. We can also use other state stores, which can be configured in the `conf/functions_worker.yml` using the field: `stateStorageProviderImplementation`, this YAML file will be parsed and loaded in Pulsar as the `WorkerConfig`.

Reviewer: ```suggestion
Currently, Pulsar uses Bookkeeper as the default state storage interface. We can also use other state stores, which can be configured in the `conf/functions_worker.yml` using the field `stateStorageProviderImplementation` (this YAML file is parsed and loaded in Pulsar as the `WorkerConfig` - see below)
```

> ## WorkerConfig

Author: @asafm I added `functions_worker.yml` related info in here

Reviewer: Looks good

Author: Then shall we restart the voting process?

The `WorkerConfig` is used to configure the Pulsar functions worker, it's parsed from a YAML file.

The `PulsarBrokerStarter`(and `PulsarStandalone`) has an optional argument `--functions-worker-conf` or `-fwc` in short to specify the path of this yaml file, if not specified, the default path: `conf/functions_worker.yml` will be used.

By using YAML as the config file for Pulsar functions worker, operators can provide complicated fields such as `List` and `Map` to the `WorkerConfig`, which is more convenient than using JSON.

Currently, the `WorkerConfig` has two fields related to the state store:

1. `stateStorageProviderImplementation`: The implementation class for the state store which should implement the interface `StateStoreProvider`, such as `org.apache.pulsar.functions.instance.state.BKStateStoreProviderImpl`

2. `stateStorageServiceUrl`: The service URL of state storage, such as: `bk://localhost:4181`

## `Runtime` and `RuntimeFactory`:

> Pulsar Function supports three kinds of runtime and, correspondingly, has three related `RuntimeFactory` to create them,

Reviewer: ```suggestion
Pulsar Function supports three kinds of runtime and, correspondingly, has three related `RuntimeFactory` to create them:
```

> <!--

Reviewer: Would it be possible to remove all the comments? There are just guidelines while you write. No need for a final PIP.

Author: sure, I will remove them

RULES
* Never place a link to an external site like Google Doc. The proposal should be in this issue entirely.
* Use a spelling and grammar checker tools if available for you (there are plenty of free ones).

PROPOSAL HEALTH CHECK
I can read the design document and understand the problem statement and what you plan to change *without* resorting to a couple of hours of code reading just to start having a high level understanding of the change.

THIS COMMENTS
Please remove them when done.
-->

> # Background knowledge

Reviewer: 💯

* Provide links where possible if a person wants to dig deeper into the background information. 

DON'T
* Do not include links *instead* explanation. Do provide links for further explanation.

EXAMPLES
* See [PIP-248](https://github.com/apache/pulsar/issues/19601), Background section to get an understanding on how you add the background knowledge needed.
  (They also included the motivation there, but ignore it as we place that in Motivation section explicitly).
-->

In Pulsar, a pulsar function support storing state, such as a `WordCount` function which stores the state of its counters.

```python
from pulsar import Function

class WordCount(Function):
    def process(self, item, context):
        for word in item.split():
            context.incr_counter(word, 1)
```

Currently, Pulsar uses Bookkeeper as the default state storage interface. We can also use other state stores, which can be configured in the `conf/functions_worker.yml` using the field: `stateStorageProviderImplementation`, this YAML file will be parsed and loaded in Pulsar as the `WorkerConfig`.

The `WorkerConfig` is used to configure the Pulsar functions worker and has two fields related to the state store:

> 1. `stateStorageProviderImplementation`: The implementation class for the state store which should implement the interface`StateStoreProvider`, such as `org.apache.pulsar.functions.instance.state.BKStateStoreProviderImpl`

Reviewer: interface`StateStoreProvider` --> interface `StateStoreProvider`

2. `stateStorageServiceUrl`: The service URL of state storage, such as: `bk://localhost:4181`

> `Runtime` and `RuntimeFactory`:

Reviewer: ```suggestion
## `Runtime` and `RuntimeFactory`
```

see:

https://github.com/apache/pulsar/blob/master/pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java#L366-L369

Then when using a custom `StateStoreProvider`, it cannot get other configurations from the `init` method.

# Goals

## In Scope

<!--
What this PIP intend to achieve once It's integrated into Pulsar.
Why does it benefit Pulsar.
-->

Make `StateStoreProvider` "truly" configurable, custom `StateStoreProvider` can be configured more easily and explicitly.

## Out of Scope

<!--
Describe what you have decided to keep out of scope, perhaps left for a different PIP/s.
-->

# High Level Design

<!--
Describe the design of your solution in *high level*.
Describe the solution end to end, from a birds-eye view.
Don't go into implementation details in this section.

I should be able to finish reading from beginning of the PIP to here (including) and understand the feature and 
how you intend to solve it, end to end.

DON'T
* Avoid code snippets, unless it's essential to explain your intent.
-->

> Make `StateStoreProvider` configurable using custom configurations.

Reviewer: After reading the detailed design, I think we can provide a more detailed yet concise description of your solution:

We will add a configuration key to the Pulsar Function Worker configuration, containing the configuration for the `StateStoreProvider` implementation as a single-line JSON String. This JSON string, a `Map<String, Object>`, will be trickled down to different runtimes,  all the way to `JavaInstanceRunnable`, the ultimate Main class all runtimes use to execute the function. This class is also the one instantiating the `StateStoreProvider`.

Now that I actually described that, I suddenly realized something.
Say the operator decided to use a different `StateStoreProvider`.
Configuring it using a single-line JSON String would not be a convenient solution for the operator.

Shouldn't we have a mechanism to configure plugins with more than just a single property?

From what I've seen, Pulsar has more than 10 different types of plugins. I bet most of them need more than one config key. Today they put it plainly as multiple different keys on the same level one. Maybe they use a shared prefix in the name.

If we suddenly pivot to something else, it ruins the consistency and provides a low operator experience. 

@codelipenghui @merlimat - what do you think?

Author: Thanks for the improvements!

And as commented above, the `stateStorageConfig` is a `Map<String, Object>` rather than a JSON string.

For a custom `StateStoreProvider`, unlike other Pulsar plugins, what fields it needs for initialization is arbitrary, so we can only provide a `Map<String, Object>` to it and make the operator decide what fields to provide in the `functions_worker.yml`.

Reviewer: Ok. I replied to this in the previous comment. Thanks

* Concrete class names and their roles and responsibility, including methods.
* Code snippets of existing code.
* Interface names and its methods.
* ...
-->

1. Add a `stateStorageConfig` to the `WorkerConfig`
2. In all three runtime factories, pass this config to the created runtime
3. Add a new cli argument to `JavaInstanceStarter` and `LocalRunner` so process&k8s runtime can pass state related config to them
4. Add new state config parameter to `JavaInstanceRunnable` to accept the state provider related configurations

## Public-facing Changes

<!--
Describe the additions you plan to make for each public facing component. 
Remove the sections you are not changing.
Clearly mark any changes which are BREAKING backward compatability.
-->

### Public API
<!--
When adding a new endpoint to the REST API, please make sure to document the following:

* path
* query parameters
* HTTP body parameters, usually as JSON.
* Response codes, and for each what they mean.
  For each response code, please include a detailed description of the response body JSON, specifying each field and what it means.
  This is the place to document the errors.
-->

### Binary protocol

### Configuration

> - A new `map<String, Object>` field like `stateStorageConfig` is added to `WorkerConfig`

Reviewer: Not exactly.
I guess you plan to add a config key named `stateStorageConfig`, but how will users enter `map<String, Object>` - don't assume they are java developers and even if they do they will don't know what to put there.

Now I presume the value type is single line JSON String?

Author: No, the `WorkerConfig` is a Java class, so this field is a `map<String, Object>` instead of a JSON string, there are already other `map<String, Object>` fields in the `WorkerConfig`, such as `functionRuntimeFactoryConfigs`, `runtimeCustomizerConfig` and `functionsWorkerServiceCustomConfigs`.

On users(operator) side, they just need to edit the `functions_worker.yml` to provide the `stateStorageConfig`, which will be parsed and loaded as the `WorkerConfig`; an example looks like the below:

```yaml
stateStorageConfig:
    username: xxx
    password: xxx
```

Author: Sure, I will add the info about `function_worker.yml` in the background

The `WorkerConfig` is used to configure the Pulsar functions worker and has two fields related to the state store:

1. `stateStorageProviderImplementation`: The implementation class for the state store which should implement the interface `StateStoreProvider`, such as `org.apache.pulsar.functions.instance.state.BKStateStoreProviderImpl`

2. `stateStorageServiceUrl`: The service URL of state storage, such as: `bk://localhost:4181`

> ## `Runtime` and `RuntimeFactory`:

Reviewer: @jiangpengcheng

> ## `Runtime` and `RuntimeFactory`:

Author: thanks for corrections, updated

# Motivation

Pulsar functions don't support configuring the `StateStoreProviderImplClass` with extra configurations, although there is a `Map<String, Object> config` parameter in the `StateStoreProvider#init` method, it's initialized as a map with only one field `stateStorageServiceUrl` in `javaInstanceRunnable`

see:

https://github.com/apache/pulsar/blob/0a39b819a9ec371322bfde61685497a2c21d9ab3/pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java#L366-L369

Then when using a custom `StateStoreProvider`, it cannot get other configurations from the `init` method.

# Goals

## In Scope

Make `StateStoreProvider` "truly" configurable, custom `StateStoreProvider` can be configured more easily and explicitly.

## Out of Scope

# High Level Design

> We will add a configuration key to the Pulsar Function Worker configuration, containing the configuration for the `StateStoreProvider` implementation as a `Map<String, Object>`.

Reviewer: I think people can't relate `map<String, object>` to the configuration of Pulsar Function Worker.  Map<string, object> is Java talk. As I replied in the comment, we need to fix this somehow.

Author: Maybe using `Map` here? As Pulsar functions use a YAML file for configuration, and YAML supports the `Map` struct

Reviewer: Yes

# High Level Design

> We will add a configuration key to the Pulsar Function Worker configuration, containing the configuration for the `StateStoreProvider` implementation as a `Map`.

Reviewer: ```suggestion
We will add a configuration key to the Pulsar Function Worker configuration, containing the configuration for the `StateStoreProvider`. Its value type will be a YAML `Map`.
```

> This `Map`, will be trickled down to different runtimes, all the way to `JavaInstanceRunnable`, the ultimate Main class all runtimes use to execute the function.

Reviewer: ```suggestion
This `Map`, will trickle down to different runtimes, all the way to `JavaInstanceRunnable`, the ultimate Main class all runtimes use to execute the function.
```
